classes/security/authorization/internal/UserAccessibleWorkflowStagePolicy.inc.php

```diff
@@ -39,11 +47,21 @@ function __construct($stageId) {
 	 */
 	function effect() {
 		$userAccessibleStages = $this->getAuthorizedContextObject(ASSOC_TYPE_ACCESSIBLE_WORKFLOW_STAGES);
-		if (empty($userAccessibleStages)) return AUTHORIZATION_DENY;
 
-		$stageId = $this->_stageId;
+		// User has no access to any stage in any workflow
+		if (empty($userAccessibleStages)) {
+			return AUTHORIZATION_DENY;
+
+		// Does user have access to this stage in the requested workflow?
+		} elseif (!is_null($this->_workflowType)) {
+			$workflowTypeRoles = Application::getWorkflowTypeRoles();
+			if (array_intersect($workflowTypeRoles[$this->_workflowType], $userAccessibleStages[$this->_stageId])) {
```

Comment:
Hmmm... is there always `$this->_stageId` key in the array `$userAccessibleStages` if the `$this->_workflowType` exists? -- Because first below and with `elseif` it will be checked `array_key_exists($this->_stageId, $userAccessibleStages)`.
Thus maybe to here do the check for AUTHORIZATION_DENY, i.e. 
```
if (!array_intersect($workflowTypeRoles[$this->_workflowType], $userAccessibleStages[$this->_stageId])) { 
return AUTHORIZATION_DENY; 
}
```
and then at the end: 
```
if (array_key_exists($this->_stageId, $userAccessibleStages)) {
return AUTHORIZATION_PERMIT;
}

```
Maybe I am wrong with the logic -- it just look so from reading the code...

Reply:
Oh, good catch. I've addressed it by checking the array key exists first:

```
if (array_key_exists($this->_stageId, $userAccessibleStages) && array_intersect($workflowTypeRoles[$this->_workflowType], $userAccessibleStages[$this->_stageId])) {
```

As a personal preference, I like to deny by default and then specify conditions for permitting. It just feels better to deny by default in case I forget certain conditions, but probably easy to mess it up either way. :)